packages/libs/contracts-sdk/contracts/facets/VincentUserViewFacet.sol

```diff
@@ -208,6 +234,77 @@ contract VincentUserViewFacet is VincentBase {
         return result;
     }
 
+    /**
+     * @notice Retrieves permitted apps for multiple PKPs with pagination
+     * @dev Takes an array of PKP token IDs and returns currently permitted apps with their version details
+     * @param pkpTokenIds Array of PKP token IDs to query
+     * @param offset The offset of the first app to retrieve for each PKP
+     * @return results Array of PkpPermittedApps structs containing permitted app data
+     */
+    function getPermittedAppsForPkps(
+        uint256[] memory pkpTokenIds, 
+        uint256 offset
+    )
+        public 
+        view 
+        returns (PkpPermittedApps[] memory results) 
+    {
+        VincentUserStorage.UserStorage storage us_ = VincentUserStorage.userStorage();
+        VincentAppStorage.AppStorage storage as_ = VincentAppStorage.appStorage();
+
+        results = new PkpPermittedApps[](pkpTokenIds.length);
+
+        for (uint256 i = 0; i < pkpTokenIds.length; i++) {
+            uint256 pkpTokenId = pkpTokenIds[i];
+
+            if (pkpTokenId == 0) {
+                revert InvalidPkpTokenId();
+            }
+
+            VincentUserStorage.AgentStorage storage agentStorage = us_.agentPkpTokenIdToAgentStorage[pkpTokenId];
+            EnumerableSet.UintSet storage permittedApps = agentStorage.permittedApps;
+            uint256 appCount = permittedApps.length();
+
+            // Count non-deleted permitted apps
+            PermittedApp[] memory tempPermittedApps = new PermittedApp[](appCount);
+            uint256 permittedCount = 0;
+
+            for (uint256 j = 0; j < appCount; j++) {
+                uint40 appId = uint40(permittedApps.at(j));
+                if (!as_.appIdToApp[appId].isDeleted) {
+                    // Get version details for the permitted app
+                    uint24 version = agentStorage.permittedAppVersion[appId];
+                    bool enabled = as_.appIdToApp[appId].appVersions[getAppVersionIndex(version)].enabled;
+                    tempPermittedApps[permittedCount] = PermittedApp({
+                        appId: appId,
+                        version: version,
+                        versionEnabled: enabled
+                    });
+                    permittedCount++;
+                }
+            }
+
+            // Apply pagination
+            uint256 start = offset;
+            uint256 end = offset + AGENT_PAGE_SIZE;
```

Comment:
I think the page size should be a param that's passed in

Reply:
The contracts sdk can also handle a default of 50 so the experience stays the same, but users have the choice to request more

Reply:
@DashKash54 no it hasn't, was waiting on confirmation - will add it